### PR TITLE
[flang] Allow lowering of sub-expressions to be overridden

### PR DESCRIPTION
**Pull request by jeanPerier**
OpenACC/OpenMP atomic lowering needs a finer control over expression lowering. This patch allows mapping evaluate::Expr<T> to mlir::Value so that any subsequent expression lowering will use these values when an operand is a mapped Expr<T>.

This is an alternative to https://github.com/llvm/llvm-project/pull/69866 From which I took the test.

The same tests as in https://github.com/llvm/llvm-project/pull/69866 are failing because the "non atomic part" is now out of the atomic.update op, which in some cases is causing verification failures because this is generated in the middle of an omp.atomic.capture. I did not try fixing these failures. My patch is about the lowering infrastructure and how to use it rather than the OpenMP semantics.

### CODE REVIEW

**Review by clementval — APPROVED** [review]
Thanks for looking at this @jeanPerier. This is a nice alternative!

**Comment by NimishMishra** [comment]
Thanks for this. In my opinion, this seems cleaner. Pinging @kiranchandramohan for his take on how to move further on this.

**Comment by kiranchandramohan** [comment]
Thanks @jeanPerier for this extension of the expression lowering infrastructure. I agree that this is much cleaner than the custom lowering. The only question I have is whether this addition of state to expression lowering will make any future parallel lowering of expressions impossible.

> Thanks for this. In my opinion, this seems cleaner. Pinging @kiranchandramohan for his take on how to move further on this.

I think you have to work with Jean on how to merge this patch. Whether the current patch can go in as it is with the failing OpenMP & OpenACC tests XFAILed, or only the expression lowering infrastructure improvement goes in and then you can update the OpenMP/OpenACC lowering code using this new infrastructure.

**Comment by kiranchandramohan** [comment]
> which in some cases is causing verification failures because this is generated in the middle of an omp.atomic.capture.

I guess the failures are because the OpenMP capture statements do not expect any statements other than `omp.atomic.read`, `omp.atomic.write` or `omp.atomic.update`. In these cases, the hoisting for the expression can go above the `omp.atomic.capture`.

**Review by razvanlupusoru — APPROVED** [review]
Yes! Thank you for taking the time to create this infrastructure! :) Definitely helps the atomic implementation!

**Comment by jeanPerier** [comment]
> The only question I have is whether this addition of state to expression lowering will make any future parallel lowering of expressions impossible.

Thanks for the review. Yes, this change does not help, but expression lowering is already depending on the converter state since it is providing the symbol map (at least when expression lowering is called via the converter).

We would first need to define the lowering parallelism requirements here. Is to lower expressions of a single statement in parallel, or is it to allow different program units to be lowered in parallel? The former seems hard and of little practicality to me. The latter makes more sense, and then, I think we could modify the bridge a bit in order to have a "converter" instance per thread/program unit lowering, in which case this change is not a blocker (sub-expression override would only impact a single thread/program unit lowering).

**Comment by jeanPerier** [comment]
> think you have to work with Jean on how to merge this patch. Whether the current patch can go in as it is with the failing OpenMP & OpenACC tests XFAILed, or only the expression lowering infrastructure improvement goes in and then you can update the OpenMP/OpenACC lowering code using this new infrastructure.

Thanks @kiranchandramohan, with your guidance to generate the non atomic expr evaluation outside of the atomic.capture. Please @razvanlupusoru, @clementval, and @NimishMishra, check that the updated tests match what you expect. Then I can push all of it if it easier.

(I rebased with hope to pass the git fetch error of the build bot)

File: flang/lib/Lower/DirectivesCommon.h [hunk 1]
```diff
@@ -269,11 +220,37 @@ static inline void genOmpAccAtomicUpdateStatement(
...
+        using T = std::decay_t<decltype(op)>;
+        if constexpr (std::is_base_of<Fortran::parser::Expr::IntrinsicBinary,
+                                      T>::value) {
+          const auto &exprLeft{std::get<0>(op.t)};
+          const auto &exprRight{std::get<1>(op.t)};
+          if (exprLeft.value().source == assignmentStmtVariable.GetSource())
+            nonAtomicSubExpr = Fortran::semantics::GetExpr(exprRight);
+          else
+            nonAtomicSubExpr = Fortran::semantics::GetExpr(exprLeft);
+        }
```
Comment (kiranchandramohan) [inline flang/lib/Lower/DirectivesCommon.h:239]: Assuming we can do something similar here for the allowed intrinsics too.

**Review by kiranchandramohan — APPROVED** [review]
> We would first need to define the lowering parallelism requirements here. Is to lower expressions of a single statement in parallel, or is it to allow different program units to be lowered in parallel? The former seems hard and of little practicality to me. The latter makes more sense, and then, I think we could modify the bridge a bit in order to have a "converter" instance per thread/program unit lowering, in which case this change is not a blocker (sub-expression override would only impact a single thread/program unit lowering).

Makes sense.

Thanks @jeanPerier once again for your help here.

**Review by NimishMishra — COMMENTED** [review]
LGTM. 

@jeanPerier  I suppose we will have to work on lowering of intrinsic procedures too. If you do not plan to extend this PR to include those, can I look into those?

File: flang/lib/Lower/DirectivesCommon.h (hunk 1 above)
Comment (jeanPerier) [reply flang/lib/Lower/DirectivesCommon.h:239]: I suppose you can indeed go through the intrinsic function arguments here too yes, @NimishMishra, I do not plan to work on this, so I am happy if you do in a later patch!

Comment (NimishMishra) [reply flang/lib/Lower/DirectivesCommon.h:239]: Thanks!

**Pull request merged**